File: app-mvp/src/main/java/com/xuie/imaginaryandroid/gui/nets/news/NewsListPresenter.java
package com.xuie.imaginaryandroid.gui.nets.news;

import com.xuie.imaginaryandroid.data.source.NetsRepository;

import io.reactivex.android.schedulers.AndroidSchedulers;

import static com.xuie.imaginaryandroid.util.Utils.checkNotNull;

/**
 * @author xuie
 * @date 17-7-5
 */
public class NewsListPresenter implements NewsListContract.Presenter {
    private NetsRepository netsRepository;
    private NewsListContract.View netsView;
    private int currentPage = 0;

    public NewsListPresenter(NetsRepository netsRepository, NewsListContract.View netsView) {
        this.netsRepository = checkNotNull(netsRepository);
        this.netsView = checkNotNull(netsView);
        this.netsView.setPresenter(this);
    }

    @Override
    public void subscribe() {
        getList(true);
    }

    @Override
    public void unsubscribe() {

    }

    @Override
    public void getList(boolean isRefresh) {
        if (isRefresh) {
            currentPage = 0;
        } else {
            currentPage += 20;
        }
        netsRepository.getNews(currentPage)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(netsSummaries -> netsView.addList(isRefresh, netsSummaries)
                        , Throwable::printStackTrace);

    }
}

File: app-mvp-kotlin/src/main/java/com/xuie/imaginaryandroid/gui/nets/news/NewsListPresenter.kt
package com.xuie.imaginaryandroid.gui.nets.news

import com.xuie.imaginaryandroid.data.source.NETSRepository

import io.reactivex.android.schedulers.AndroidSchedulers

/**
 * Created by xuie on 17-7-5.
 */

class NewsListPresenter(
        private val netsRepository: NETSRepository,
        private val netsView: NewsListContract.View
) : NewsListContract.Presenter {
    private var currentPage = 0

    init {
        this.netsView.presenter = this
    }

    override fun subscribe() {
        getList(true)
    }

    override fun unsubscribe() {

    }

    override fun getList(isRefresh: Boolean) {
        if (isRefresh) {
            currentPage = 0
        } else {
            currentPage += 20
        }
        netsRepository.getNews(currentPage)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ netsSummaries ->
                    netsView.addList(isRefresh, netsSummaries) }, { e -> e.printStackTrace()})

    }
}
